core/fedavgoptimizer.py

```diff
@@ -0,0 +1,144 @@
+from core.utils.event_types import ActionableEventTypes, RawEventTypes, callback_handler
+from core.utils.dmljob import deserialize_job
+from core.utils.dmlresult import DMLResult
+
+class FederatedAveragingOptimizer(object):
+	"""
+	Federated Averaging Optimizer
+
+	Implements the Federated Averaging algorithm (with some tweaks)
+	and acts as a "brain" for the Unix Service
+
+	"""
+	def __init__(self, content):
+		"""
+		Kwargs is a dictionary containing all the info needed to initialize the Optimizer.
+		This includes any number of properties such as thresholds, etc.
+		"""
+		kwargs = content.get('optimizer_params')
+		serialized_job = content.get('serialized_job')
+		# The Optimizer knows its job
+		self.job = deserialize_job(serialized_job)
+		self.listen_iterations = kwargs.get('listen_iterations', 0)
+		self.listen_bound = kwargs.get('listen_bound', 1)
+		self.LEVEL_1_CALLBACKS = {
```

Comment:
What are level 1, 2, and 3 callbacks? Please write some docstring about them.

Reply:
done